doc/perception_pipeline/perception_pipeline_tutorial.rst

```diff
@@ -160,9 +164,10 @@ After running the code, you should be able to see something like this in RViz:
 
 Running the Code
 ++++++++++++++++
-Roslaunch the launch file to run the code directly from moveit_tutorials: ::
 
- roslaunch moveit_tutorials detect_and_add_cylinder_collision_object_demo.launch
+Keep the launch file from above running and run the code directly from moveit_tutorials: ::
+
+ rosrun moveit_tutorials detect_and_add_cylinder_collision_object_demo
 
 KNOWN ISSUE - You may see the following error when running the demo ::
```

Comment:
Does this PR address this issue, enough to remove this disclaimer? Below the fold it says "We are working on fixing it" which is not true, no one has commented on the linked issue in 3 years

Reply:
The source of the error is that the shape updater used for self-filtering has to wait for transforms from the whole robot to the planning_frame.
The waiting time can be configured by a magic ROS parameter and [I increased the value to 0.5 seconds](https://github.com/ros-planning/moveit_tutorials/pull/633/commits/210ac1b2572d204bef01120d2085095ba9bb3a5c).
That works well on my intel i5 laptop and I expected it would be enough time for the `robot_state_publisher` to publish the robot transforms even on a weak user machine.
@JStech Could you please try different values and see how much time is actually needed on your system to get rid of these errors?
Alternatively (or in addition) we could throttle the publish frequency of the point cloud and reduce the maximum iterations of the RANSAC runs to reduce some computational load.

Reply:
This was sufficient to stop the errors for me:
``` diff
diff --git a/doc/perception_pipeline/src/bag_publisher_maintain_time.cpp b/doc/perception_pipeline/src/bag_publisher_maintain_time.cpp
index fa40247..1ad1b96 100644
--- a/doc/perception_pipeline/src/bag_publisher_maintain_time.cpp
+++ b/doc/perception_pipeline/src/bag_publisher_maintain_time.cpp
@@ -71,7 +71,7 @@ int main(int argc, char** argv)
   ros::Rate loop_rate(0.2);
   while (ros::ok())
   {
-    point_cloud_ptr->header.stamp = ros::Time::now();
+    point_cloud_ptr->header.stamp = ros::Time::now() - ros::Duration(0.1);
     point_cloud_publisher.publish(*point_cloud_ptr);
     loop_rate.sleep();
   }
```

Reply:
And increasing `shape_transform_cache_lookup_wait_time` to 0.6 seconds instead does not work?
I'm not happy with introducing a delay to the "camera driver".
If the driver is that perfect and can provide current data on the topic, MoveIt should be able to handle that, right?

Reply:
Sorry, I had already tried this and came here to post in when I saw your request, above. I've not tried changing `shape_transform_cache_lookup_wait_time`.

Reply:
@JStech could you please try again with an even higher `lookup_wait_time`?
I don't want to use your patch unless there is a sound justification for why we should not publish the *current* point cloud.

Reply:
Yes, it still happens. I increased it to 30.0 and it just delayed when the error was printed (see timestamps in console log below). I don't think that parameter actually has any effect on this error.
```
[ INFO] [1620749352.107751040]: Listening to '/move_group/monitored_planning_scene'
[ERROR] [1620749381.518885835]: Transform error: Lookup would require extrapolation into the past.  Requested time 1620749351.466759904 but the earliest data is at time 1620749371.528618097, when looking up transform from frame [panda_hand] to frame [camera_rgb_optical_frame]
[ERROR] [1620749381.518986350]: Transform cache was not updated. Self-filtering may fail. If transforms were not available yet, consider setting robot_description_planning/shape_transform_cache_lookup_wait_time to wait longer for transforms
[ERROR] [1620749411.594371880]: Transform error: Lookup would require extrapolation into the past.  Requested time 1620749361.468870542 but the earliest data is at time 1620749401.628633975, when looking up transform from frame [panda_hand] to frame [camera_rgb_optical_frame]
[ERROR] [1620749411.594566074]: Transform cache was not updated. Self-filtering may fail. If transforms were not available yet, consider setting robot_description_planning/shape_transform_cache_lookup_wait_time to wait longer for transforms
[ERROR] [1620749441.614890057]: Transform error: Lookup would require extrapolation into the past.  Requested time 1620749366.466824576 but the earliest data is at time 1620749431.528650045, when looking up transform from frame [panda_hand] to frame [camera_rgb_optical_frame]
[ERROR] [1620749441.615061107]: Transform cache was not updated. Self-filtering may fail. If transforms were not available yet, consider setting robot_description_planning/shape_transform_cache_lookup_wait_time to wait longer for transforms
[ERROR] [1620749471.643072340]: Transform error: Lookup would require extrapolation into the past.  Requested time 1620749371.466826179 but the earliest data is at time 1620749461.728646993, when looking up transform from frame [panda_hand] to frame [camera_rgb_optical_frame]
[ERROR] [1620749471.643152101]: Transform cache was not updated. Self-filtering may fail. If transforms were not available yet, consider setting robot_description_planning/shape_transform_cache_lookup_wait_time to wait longer for transforms
```

Reply:
I just took another look at this. These errors are about extrapolating into the *past*, so obviously the wait time won't fix that (my workaround didn't fix it either).

What *did* fix it was adding a half second sleep before the publish loop.

Reply:
Does that fix it reliably?

Reply:
Sorry for leaving this hanging out to dry for so long, I'm overwhelmed by all the upcoming things to do on MoveIt and my personal availability.

> What did fix it was adding a half second sleep before the publish loop.

I'm somewhat unsure what you mean here. Do you actually mean [before the loop](https://github.com/ros-planning/moveit_tutorials/blob/36f4bd6d950de8340a72edba568da5904d81d957/doc/perception_pipeline/src/bag_publisher_maintain_time.cpp#L72) or [before the `publish` call](https://github.com/ros-planning/moveit_tutorials/blob/36f4bd6d950de8340a72edba568da5904d81d957/doc/perception_pipeline/src/bag_publisher_maintain_time.cpp#L75).
I just made the publisher latched (makes sense to me anyway with a 0.2s publish rate and increased the queue size.
If you meant the first location, this should solve it as well.

I still wouldn't be happy to add a sleep to the second location because it artificially delays messages and if MoveIt can't handle perfect times we should fix that in MoveIt.
However, by now I'm more interested in finally getting this merged, so if you say the problem only disappears with a sleep before publishing, I'd add it for now with a note.

Please confirm @JStech .

Reply:
ping @JStech

Reply:
Sorry, been busy here too. I mean before the loop:
```
  ros::Rate loop_rate(0.2);
  ros::Duration(0.5).sleep();
  while (ros::ok())
  ...
```

Reply:
I also just pulled, built, and ran the latest version (cab19fd), and I still get the error. (And the delay still fixes it.)